### PR TITLE
pcre: use cmake config files instead of meson

### PR DESCRIPTION
**Pull request by neheb**
Less error prone.

Signed-off-by: Rosen Penev <rosenp@gmail.com>

### CODE REVIEW

File: subprojects/packagefiles/pcre/meson.build [hunk 1]
```diff
@@ -80,24 +149,22 @@ char_tables = custom_target('chartables',
...
+  link_with: pcre8,
+)
 
 pcretest = executable('pcretest',
   'pcretest.c', 'pcre_printint.c',
-  c_args : compile_args,
-  link_with : [pcre8, pcre_posix],
-  include_directories : conf_inc,
+  c_args: compile_args,
+  dependencies: libedit_dep,
```
Comment (xclaesse) [inline subprojects/packagefiles/pcre/meson.build:166]: Does it build if libedit_dep is not found?

File: subprojects/packagefiles/pcre/meson.build [hunk 2]
```diff
@@ -80,24 +149,22 @@ char_tables = custom_target('chartables',
   output : 'pcre_chartables.c',
   command : [dftables, '@OUTPUT@'])
 
-pcre8 = build_target('pcre',
+pcre8 = library('pcre',
   pcre8_sources, char_tables,
-  c_args : compile_args,
-  target_type : libtype,
-  include_directories : conf_inc)
+  c_args: compile_args,
```
Comment (xclaesse) [inline subprojects/packagefiles/pcre/meson.build:154]: Shouldn't you pass all deps you lookup above?

File: subprojects/packagefiles/pcre/meson.build (hunk 1 above)
Comment (neheb) [reply subprojects/packagefiles/pcre/meson.build:166]: It does yes.

File: subprojects/packagefiles/pcre/meson.build (hunk 2 above)
Comment (neheb) [reply subprojects/packagefiles/pcre/meson.build:154]: Yeah probably. The interesting thing is that these dependencies seem to be used not by the library but utilities that it comes with.

Comment (xclaesse) [reply subprojects/packagefiles/pcre/meson.build:154]: Do we build those utilities? If not, then we don't need to lookup for those deps at all.

Comment (neheb) [reply subprojects/packagefiles/pcre/meson.build:154]: pcretest is built, yes.

I guess since they're all optional they can be removed.

Comment (neheb) [reply subprojects/packagefiles/pcre/meson.build:154]: Yeah:

```
~/p/pcre-8.45 (master)> git grep SUPPORT_LIBZ *.c
pcregrep.c:#ifdef SUPPORT_LIBZ
pcregrep.c:#ifdef SUPPORT_LIBZ
pcregrep.c:#if defined SUPPORT_LIBZ || defined SUPPORT_LIBBZ2
pcregrep.c:#ifdef SUPPORT_LIBZ
pcregrep.c:#ifdef SUPPORT_LIBZ
pcregrep.c:#ifdef SUPPORT_LIBZ
pcregrep.c:#ifdef SUPPORT_LIBZ
pcregrep.c:#if defined SUPPORT_LIBZ || defined SUPPORT_LIBBZ2
pcregrep.c:#if defined SUPPORT_LIBZ || defined SUPPORT_LIBBZ2
pcregrep.c:#ifdef SUPPORT_LIBZ
pcregrep.c:#ifdef SUPPORT_LIBZ
~/p/pcre-8.45 (master)> git grep SUPPORT_LIBB *.c
pcregrep.c:#ifdef SUPPORT_LIBBZ2
pcregrep.c:#ifdef SUPPORT_LIBBZ2
pcregrep.c:#if defined SUPPORT_LIBZ || defined SUPPORT_LIBBZ2
pcregrep.c:#ifdef SUPPORT_LIBBZ2
pcregrep.c:#ifdef SUPPORT_LIBBZ2
pcregrep.c:#ifdef SUPPORT_LIBBZ2
pcregrep.c:#ifdef SUPPORT_LIBBZ2
pcregrep.c:#if defined SUPPORT_LIBZ || defined SUPPORT_LIBBZ2
pcregrep.c:#if defined SUPPORT_LIBZ || defined SUPPORT_LIBBZ2
pcregrep.c:#ifdef SUPPORT_LIBBZ2
pcregrep.c:#ifdef SUPPORT_LIBBZ2
pcregrep.c:#ifdef SUPPORT_LIBBZ2
~/p/pcre-8.45 (master)> git grep SUPPORT_LIBE *.c
pcretest.c:#if defined(SUPPORT_LIBREADLINE) || defined(SUPPORT_LIBEDIT)
pcretest.c:#if defined(SUPPORT_LIBREADLINE) || defined(SUPPORT_LIBEDIT)
pcretest.c:#if defined(SUPPORT_LIBREADLINE) || defined(SUPPORT_LIBEDIT)
mangix@TrapNAS ~/p/pcre-8.45 (master)> git grep SUPPORT_LIBR *.c
pcretest.c:#if defined(SUPPORT_LIBREADLINE) || defined(SUPPORT_LIBEDIT)
pcretest.c:#if defined(SUPPORT_LIBREADLINE)
pcretest.c:#if defined(SUPPORT_LIBREADLINE) || defined(SUPPORT_LIBEDIT)
pcretest.c:#if defined(SUPPORT_LIBREADLINE) || defined(SUPPORT_LIBEDIT)
```

It's all pcretest and pcregrep. The former is built but not the latter.

Comment (xclaesse) [reply subprojects/packagefiles/pcre/meson.build:152]: ok good

File: subprojects/packagefiles/pcre/meson_options.txt [hunk 3]
```diff
@@ -1 +1,7 @@
-option('shared_lib', type : 'boolean', value : false)
```
Comment (eli-schwartz) [inline subprojects/packagefiles/pcre/meson_options.txt:1]: Not strictly part of the stated purpose of this PR.

On the other hand, thank you! :D :D :D

Comment (neheb) [reply subprojects/packagefiles/pcre/meson_options.txt:1]: Yeah that's probably just a cmake option to deal with cmake's inability to have both library types. At least not easily.

**Comment by xclaesse** [comment]
@neheb I don't understand why, but this PR breaks pcre symbol visibility on Windows, all wraps that depend on pcre fails because of this.

```
2022-08-28T00:42:51.5733561Z [207/1248] "link" @subprojects/glib-2.72.2/glib/glib-2.0-0.dll.rsp
2022-08-28T00:42:51.5735299Z FAILED: subprojects/glib-2.72.2/glib/glib-2.0-0.dll subprojects/glib-2.72.2/glib/glib-2.0-0.pdb 
2022-08-28T00:42:51.5815967Z "link" @subprojects/glib-2.72.2/glib/glib-2.0-0.dll.rsp
2022-08-28T00:42:51.5821853Z    Creating library subprojects\glib-2.72.2\glib\glib-2.0.lib and object subprojects\glib-2.72.2\glib\glib-2.0.exp
2022-08-28T00:42:51.5824645Z D__a_wrapdb_wrapdb_subprojects_glib-2.72.2_glib_gregex.c.obj : error LNK2001: unresolved external symbol pcre_free```

**Comment by neheb** [comment]
I'll look into it.

**Comment by eli-schwartz** [comment]
This PR has a nonsense change unrelated to the commit message.

That change should not have passed review, it got rid of a highly intentional Windows workaround.

**Comment by xclaesse** [comment]
> it got rid of a highly intentional Windows workaround

Could you please elaborate on what was intentional?

**Comment by xclaesse** [comment]
I tried to debug myself and I haven't found why dll are not working. I'm wondering if shared lib ever worked even upstream. We were building static libraries by default before this PR, that could be why it ever worked.

**Comment by eli-schwartz** [comment]
Literally that.

Dlls didn't work on Windows, the previous version of this wrap acknowledged that and forced a static library.

This PR ceased that, and it did so silently, without stating in the commit message that it was doing so. This is wrong for several reasons, but the main reason that matters today is that by hiding what it did, it failed to draw attention to a reviewable change, that change did not in fact get reviewed, and that change created a regression.

**Comment by xclaesse** [comment]
Ok, I did not know that whole story. I assumed since it was using build_target() is was just decade old code. Thanks for the clarification.

**Comment by xclaesse** [comment]
Side note, glib master has been switched to pcre2, that will be part of next stable release 2.74. So this pcre1 wrap will become less important to me at least.

**Comment by xclaesse** [comment]
https://github.com/mesonbuild/wrapdb/pull/608

**Comment by eli-schwartz** [comment]
build_target isn't deprecated, it's highly useful in modern code, as long as that code needs to set the library type based on something like the platform instead of or in addition to user options.

I knew I'd seen it used that way before, but mea culpa, pcre wasn't one of them.

It's still a suspicious change of the previous wrap version -- before, it defaulted to static, now it defaults to shared.

But it's not obvious why, like it's obvious elsewhere.

**Pull request merged**